### PR TITLE
Is/feature/custom headers

### PR DESCRIPTION
**Pull request by isaacakakpo1**
[WEBRTC-4632 - Ticket title](https://telnyx.atlassian.net/browse/TELAPPS-4632)
---

Passing and Accepting Custom Headers 
-  passing custom headers from SIP INVITE to RTC INVITE (session_id=dcc9f896-6821-11ee-9e86-02420a0d3a69):

 - passing custom headers from RTC INVITE to SIP INVITE (session_id=cc59cbec-6823-11ee-acca-02420a0d6669)
it works. 

- passing custom headers from answer RTC message on SIP responses (session_id=3e734902-6823-11ee-895c-02420a0d3a69)

### After changes
Custom headers are passed for INVITE and ANSWER Messages

## TODO
Complete Tutorial Docs.

## ✋ Manual testing
All call functionality works

### CODE REVIEW

File: TelnyxRTC/Telnyx/TxClient.swift [hunk 1]
```diff
@@ -410,7 +435,7 @@ extension TxClient {
                         ringtone: self.txConfig?.ringtone,
                         ringbackTone: self.txConfig?.ringBackTone,
                         iceServers: self.serverConfiguration.webRTCIceServers)
-        call.newCall(callerName: callerName, callerNumber: callerNumber, destinationNumber: destinationNumber, clientState: clientState)
+        call.newCall(callerName: callerName, callerNumber: callerNumber, destinationNumber: destinationNumber, clientState: clientState,customHeaders: customHeaders)
```
Comment (RJSkorski) [inline TelnyxRTC/Telnyx/TxClient.swift:438]: add space prior to `custom_headers`

File: TelnyxRTC/Telnyx/TxClient.swift [hunk 2]
```diff
@@ -663,7 +689,7 @@ extension TxClient : SocketDelegate {
                let callUUIDString = params["callID"] as? String,
                let callUUID = UUID(uuidString: callUUIDString),
                let call = calls[callUUID] {
-                call.handleVertoMessage(message: vertoMessage,txclient: self)
+                call.handleVertoMessage(message: vertoMessage,dataMessage: message,txClient: self)
```
Comment (RJSkorski) [inline TelnyxRTC/Telnyx/TxClient.swift:692]: another space prior to dataMessage

File: TelnyxRTC/Telnyx/Verto/AnswerMessage.swift [hunk 3]
```diff
@@ -26,6 +28,15 @@ class AnswerMessage : Message {
 
         params["sessionId"] = sessionId
         params["sdp"] = sdp
+        if(!customHeaders.isEmpty){
```
Comment (RJSkorski) [inline TelnyxRTC/Telnyx/Verto/AnswerMessage.swift:31]: I see that code in 2 places, shall we extract it in a separate function?

File: TelnyxRTC/Telnyx/WebRTC/Call.swift [hunk 4]
```diff
@@ -213,14 +220,14 @@ public class Call {
         It sets the incoming sdp as the remoteDecription.
         sdp: Is the remote SDP to configure in the current RTCPeerConnection
      */
-    private func answered(sdp: String) {
+    private func answered(sdp: String,custumHeaders:[String:String] = [:]) {
```
Comment (RJSkorski) [inline TelnyxRTC/Telnyx/WebRTC/Call.swift:223]: please add a space

File: TelnyxRTC/Telnyx/WebRTC/Call.swift [hunk 5]
```diff
@@ -275,12 +282,13 @@ extension Call {
...
                  destinationNumber: String,
-                 clientState: String? = nil) {
+                          clientState: String? = nil,
+                          customHeaders:[String:String] = [:]) {
         if (destinationNumber.isEmpty) {
             Logger.log.e(message: "Call:: Please enter a destination number.")
             return
         }
-        invite(callerName: callerName, callerNumber: callerNumber, destinationNumber: destinationNumber, clientState: clientState)
+        invite(callerName: callerName, callerNumber: callerNumber, destinationNumber: destinationNumber, clientState: clientState,customHeaders: customHeaders)
```
Comment (RJSkorski) [inline TelnyxRTC/Telnyx/WebRTC/Call.swift:291]: please add a space

File: TelnyxRTC/Telnyx/WebRTC/Call.swift [hunk 6]
```diff
@@ -472,14 +487,13 @@ extension Call : PeerDelegate {
  */
 extension Call {
 
-    internal func handleVertoMessage(message: Message,txclient:TxClient) {
+    internal func handleVertoMessage(message: Message,dataMessage: String,txClient:TxClient) {
```
Comment (RJSkorski) [inline TelnyxRTC/Telnyx/WebRTC/Call.swift:490]: please add a space

File: TelnyxRTC/Telnyx/WebRTC/Call.swift [hunk 7]
```diff
@@ -509,8 +523,21 @@ extension Call {
                 } else {
                     Logger.log.w(message: "Call:: .ANSWER missing SDP")
                 }
+                var customHeaders = [String:String]()
+                if params["dialogParams"] is [String:Any] {
```
Comment (RJSkorski) [inline TelnyxRTC/Telnyx/WebRTC/Call.swift:527]: This code is in 2 places as well. can we extract it into a separate function?

File: TelnyxWebRTCDemo/Extensions/AppDelegateCallKitExtension.swift [hunk 8]
```diff
@@ -156,7 +156,7 @@ extension AppDelegate : CXProviderDelegate {
     func provider(_ provider: CXProvider, perform action: CXAnswerCallAction) {
         print("AppDelegate:: ANSWER call action: callKitUUID [\(String(describing: self.callKitUUID))] action [\(action.callUUID)]")
 
-        self.telnyxClient?.answerFromCallkit(answerAction: action)
+        self.telnyxClient?.answerFromCallkit(answerAction: action,customHeaders:  ["X-test-answer":"ios-test"])
```
Comment (RJSkorski) [inline TelnyxWebRTCDemo/Extensions/AppDelegateCallKitExtension.swift:159]: Add space

File: TelnyxWebRTCDemo/Extensions/ViewControllerVoIPExtension.swift [hunk 9]
```diff
@@ -127,9 +134,10 @@ extension ViewController : VoIPDelegate {
             }
             
             let call = try telnyxClient?.newCall(callerName: callerName,
-                                                         callerNumber: callerNumber,
-                                                         destinationNumber: destinationNumber,
-                                                         callId: callUUID)
+                                                 callerNumber: callerNumber,
+                                                 destinationNumber: destinationNumber,
+                                                 callId: callUUID,customHeaders: ["X-test1":"ios-test1",
```
Comment (RJSkorski) [inline TelnyxWebRTCDemo/Extensions/ViewControllerVoIPExtension.swift:139]: should we move the last param to the new line?

File: TelnyxRTC/Telnyx/TxClient.swift [hunk 10]
```diff
@@ -165,6 +166,20 @@ public class TxClient {
             RTCAudioSession.sharedInstance().isAudioEnabled = newValue
         }
     }
+    
+    let currentRoute = AVAudioSession.sharedInstance().currentRoute
+    
+    func isSppeakerEnabled() -> Bool {
```
Comment (RJSkorski) [inline TelnyxRTC/Telnyx/TxClient.swift:172]: misspelling: double p in speaker

File: TelnyxRTC/Telnyx/TxClient.swift [hunk 11]
```diff
@@ -228,14 +243,21 @@ public class TxClient {
     }
     
     /// To answer and control callKit active flow
-    public func answerFromCallkit(answerAction:CXAnswerCallAction) {
+    /// - Parameters:
+    ///     - answerAction : `CXAnswerCallAction` from callKit
+    ///     - customHeaders: (Optional)
+    public func answerFromCallkit(answerAction:CXAnswerCallAction,customHeaders:[String:String] = [:]) {
```
Comment (RJSkorski) [inline TelnyxRTC/Telnyx/TxClient.swift:249]: add a space

**Review by RJSkorski — APPROVED** [review]

File: TelnyxRTC/Telnyx/Verto/AnswerMessage.swift (hunk 3 above)
Comment (isaacakakpo1) [reply TelnyxRTC/Telnyx/Verto/AnswerMessage.swift:30]: I've refactored it but It'll good to have `customHeaders.isEmpty` , so we only pass if necessary

**Review by RJSkorski — APPROVED** [review]

**Pull request merged**